vimms/MassSpec.py

```diff
@@ -102,6 +102,101 @@ def __init__(self, chem, query_rt, ms_level, peaks, scan_id):
     def __repr__(self):
         return 'MS%d FragmentationEvent for %s at %f' % (self.ms_level, self.chem, self.query_rt)
 
+
+class TaskManager(object):
```

Comment:
I'm not sure if this class is thread-safe. If it's not, are we sure that there will only ever by one thread acting on it (and internal data being manipulated externally, e.g. pending_tasks).

Reply:
In the simulator, there's currently no threading at all, so it's fine.

On IAPI, the codes are certainly not thread-safe.

When running on the real mass spec, all the scan handing codes (including calls to this `TaskManager`) happen inside the event handler for MsScanArrived event from IAPI. I can't find any documentation on how event handling is called from Pythonnet, but I think we could assume that each handler is called in its own thread. 

Therefore it is possible for two scans to arrive at such close intervals that we have both event handling threads running at the same time, reading and writing to the same data structure at the same time, and bad things could happen. In fact, this has been an issue with our codes all along -- even without this `TaskManager` thing, e.g. [see here](https://github.com/sdrogers/vimms-fusion/blob/master/vimms_fusion/MassSpec.py#L238-L240). 

The only reason we never hit any problem with this so far is because only the MS1 scans are processed for computations, while MS2 scans are simply received and stored. The gaps between successive MS1 scans are big enough that we never encountered the problem with thread-safety. But still, it's good to fix this now before it becomes a problem. I've added [an issue](https://github.com/sdrogers/vimms/issues/201) for this.